src/analysis/normed_space/M_structure.lean

```diff
@@ -0,0 +1,293 @@
+
+-/
+
+variables {X : Type*} [normed_group X]
+
+variables {𝕜 : Type*} [normed_field 𝕜] [normed_space 𝕜 X]
+
+/--
+A projection on a normed space `X` is said to be an L-projection if, for all `x` in `X`,
+$$
+∥x∥ = ∥P x∥ + ∥(1 - P) x∥.
+$$
+-/
+structure is_Lprojection (P : X →L[𝕜] X) : Prop :=
+(proj : is_idempotent_elem P)
+(Lnorm : ∀ (x : X), ∥x∥ = ∥P x∥ + ∥(1 - P) x∥)
+
+/--
+A projection on a normed space `X` is said to be an M-projection if, for all `x` in `X`,
+$$
+∥x∥ = max(∥P x∥, ∥(1 - P) x∥).
+$$
+-/
+structure is_Mprojection (P : X →L[𝕜] X) : Prop :=
+(proj : is_idempotent_elem P)
+(Mnorm : ∀ (x : X), ∥x∥ = (max ∥P x∥  ∥(1 - P) x∥))
+
+namespace is_Lprojection
+
+lemma Lcomplement {P : X →L[𝕜] X} (h: is_Lprojection P) :  is_Lprojection (1 - P) :=
+⟨h.proj.one_sub, λ x, by { rw [add_comm, sub_sub_cancel], exact h.Lnorm x }⟩
+
+lemma Lcomplement_iff (P : X →L[𝕜] X) : is_Lprojection P ↔ is_Lprojection (1 - P) :=
+⟨Lcomplement, λ h, sub_sub_cancel 1 P ▸ h.Lcomplement⟩
+
+lemma commute {P Q : X →L[𝕜] X} (h₁ : is_Lprojection P) (h₂ : is_Lprojection Q) : commute P Q :=
+begin
+  have PR_eq_RPR : ∀ R : (X →L[𝕜] X), is_Lprojection R →  (P * R = R * P * R) :=λ R h₃,
+  begin
+    ext,
+    rw ← norm_sub_eq_zero_iff,
+    have e1 : ∥R x∥ ≥ ∥R x∥ + 2 • ∥ (P * R) x - (R * P * R) x∥ :=
+    calc ∥R x∥ = ∥R (P (R x))∥ + ∥(1 - R) (P (R x))∥ + (∥(R * R) x - R (P (R x))∥
+      + ∥(1 - R) ((1 - P) (R x))∥) :
+      by rw [h₁.Lnorm, h₃.Lnorm, h₃.Lnorm ((1 - P) (R x)), continuous_linear_map.sub_apply 1 P,
+        continuous_linear_map.one_apply, map_sub, continuous_linear_map.coe_mul]
+    ... = ∥R (P (R x))∥ + ∥(1 - R) (P (R x))∥ + (∥R x - R (P (R x))∥
+      + ∥((1 - R) * R) x - (1 - R) (P (R x))∥) : by rw [h₃.proj.eq,
+        continuous_linear_map.sub_apply 1 P, continuous_linear_map.one_apply,
+        map_sub,continuous_linear_map.coe_mul]
+    ... = ∥R (P (R x))∥ + ∥(1 - R) (P (R x))∥ + (∥R x - R (P (R x))∥ + ∥(1 - R) (P (R x))∥) :
+      by rw [sub_mul, h₃.proj.eq, one_mul, sub_self, continuous_linear_map.zero_apply, zero_sub,
+        norm_neg]
+    ... = ∥R (P (R x))∥ + ∥R x - R (P (R x))∥ + 2•∥(1 - R) (P (R x))∥  : by abel
+    ... ≥ ∥R x∥ + 2 • ∥ (P * R) x - (R * P * R) x∥ :
+      by exact add_le_add_right (norm_le_insert' (R x) (R (P (R x)))) (2•∥(1 - R) (P (R x))∥),
+    rw ge at e1,
+    nth_rewrite_rhs 0 ← add_zero (∥R x∥) at e1,
+    rw [add_le_add_iff_left, two_smul,  ← two_mul]  at e1,
+    rw le_antisymm_iff,
+    refine ⟨_, norm_nonneg _⟩,
+    rwa [←mul_zero (2 : ℝ), mul_le_mul_left (show (0 : ℝ) < 2, by norm_num)] at e1
+  end,
+  have QP_eq_QPQ : Q * P = Q * P * Q :=
+  begin
+    have e1 : P * (1 - Q) = P * (1 - Q) - (Q * P - Q * P * Q) :=
+    calc P * (1 - Q) = (1 - Q) * P * (1 - Q) : by rw PR_eq_RPR (1 - Q) h₂.Lcomplement
+    ... = P * (1 - Q) - (Q * P - Q * P * Q) : by noncomm_ring,
+    rwa [eq_sub_iff_add_eq, add_right_eq_self, sub_eq_zero] at e1
+  end,
```

Comment:
we normally just put this in a tactic block, not `begin` and `end`. The same goes for the previous `have`.
```suggestion
  have QP_eq_QPQ : Q * P = Q * P * Q,
  { have e1 : P * (1 - Q) = P * (1 - Q) - (Q * P - Q * P * Q) :=
    calc P * (1 - Q) = (1 - Q) * P * (1 - Q) : by rw PR_eq_RPR (1 - Q) h₂.Lcomplement
    ... = P * (1 - Q) - (Q * P - Q * P * Q) : by noncomm_ring,
    rwa [eq_sub_iff_add_eq, add_right_eq_self, sub_eq_zero] at e1 },
```